acceptance/terraform: fix nodectl

cuongdo opened the pull request:
Fixes `TestRebalance_3To5Small` failures from last 2 nights

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10628)
<!-- Reviewable:end -->

Commit:
acceptance/terraform: ease debugging

Allow passing in of the node index (an integer >= 1) through the
NODE_INDEX environment variable. This allows testing of `nodectl` on
`gceworker`s as follows:

```
NODE_INDEX=1 ./nodectl download gs://cockroach-test/allocatortest/3nodes-10g-262ranges
```

Commit:
acceptance/terraform: fix nodectl failures

When downloading archived stores, `nodectl` was refusing to do work
when .sst files already existed in the destination. This causes a race
with our allocator Terraform tests, which start nodes briefly (to fail
faster if there's a basic test issue) then run `nodectl download`. If
CockroachDB writes data fast enough that .sst files are created,
`nodectl` can fail on this check.

This fixes the race by backing up old files in the cockroach data
directory. Backing up is a convenience measure for anyone testing
`nodectl upload` and `nodectl download` on the same machine.

tamird commented:
Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/terraform/nodectl, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10628#-KWIvwNmaT-c5v41nMpE:-KWIvwNmaT-c5v41nMpF:b-mfu9vn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f1c27428d4171134fc82dd76fb2c2725adaee80/pkg/acceptance/terraform/nodectl#L29)):*
> ```
> gcs_url="$2"
> hostname="$(hostname)"
> if [ -z ${NODE_INDEX+x} ]; then
> ```

I believe this can be written `node_index=${NODE_INDEX-${hostname##*-}}`. Tested locally.

---

*[pkg/acceptance/terraform/nodectl, line 18 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10628#-KWIxY8shC3w5iKUGT_-:-KWIxY8shC3w5iKUGT_0:brsly18) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8f0b1becedbd336a9440e186f596e384afb3adf/pkg/acceptance/terraform/nodectl#L18)):*
> ```
> # Backup all regular files in the data directory.
> backup_data_dir() {
>     local backup_dir="${DATA_DIR}/backup-$(date +%Y%m%d-%H%M%S)"
> ```

this produces the following warning in shellcheck: `^-- SC2155: Declare and assign separately to avoid masking return values.`

might be worth running shellcheck over this script, or at least the parts you're adding.

---

*[pkg/acceptance/terraform/nodectl, line 21 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10628#-KWIxfO0ZEN5WYq8e4qG:-KWIxfO0ZEN5WYq8e4qH:b-wvwr7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8f0b1becedbd336a9440e186f596e384afb3adf/pkg/acceptance/terraform/nodectl#L21)):*
> ```
>     echo "Backing up files in ${DATA_DIR} to ${backup_dir}"
>     mkdir "${backup_dir}"
>     for f in "${DATA_DIR}"/*; do
> ```

perhaps `find "${DATA_DIR}" -type f | xargs -J % mv % "$backup_dir"`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10628)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/terraform/nodectl, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10628#-KWIvwNmaT-c5v41nMpE:-KWKNP3lvCk-sU-Eym0q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f1c27428d4171134fc82dd76fb2c2725adaee80/pkg/acceptance/terraform/nodectl#L29)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I believe this can be written `node_index=${NODE_INDEX-${hostname##*-}}`. Tested locally.
</details>
Done.

---

*[pkg/acceptance/terraform/nodectl, line 18 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10628#-KWIxY8shC3w5iKUGT_-:-KWKN5OHbhq92hYCa46c:bz5u2o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8f0b1becedbd336a9440e186f596e384afb3adf/pkg/acceptance/terraform/nodectl#L18)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this produces the following warning in shellcheck: `^-- SC2155: Declare and assign separately to avoid masking return values.`
> 
> might be worth running shellcheck over this script, or at least the parts you're adding.
</details>
Done.

I've run `shellcheck` over this script, and I've fixed all but one issue (which is intentional).

---

*[pkg/acceptance/terraform/nodectl, line 21 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10628#-KWIxfO0ZEN5WYq8e4qG:-KWKOoUED6OCCYuLlGFa:bwmaidt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8f0b1becedbd336a9440e186f596e384afb3adf/pkg/acceptance/terraform/nodectl#L21)):*
> find "${DATA_DIR}" -type f | xargs -J % mv % "$backup_dir"

Good suggestion, though `-J` on macOS seems to map to `-I` on Linux. I changed this to a more complicated `find` command, which I hope is a bit more cross-platform.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10628)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10628#-:-KWKWRrsk9uFl7CYqv0D:bnfp4nl)*
<!-- Sent from Reviewable.io -->
